W-21210749 split model policies gr 2 (#994)
mulesoft/docs-gateway: pdk/1.7/modules/ROOT/pages/policies-pdk-create-project-split.adoc

= Creating a Split-Model Policy Project
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

To begin developing a new Split-Model custom policy with Flex Gateway Policy Development Kit (PDK), generate a new custom policy project.

> To create a Split-Model project, complete the following tasks:

Reviewer: ```suggestion
To create a Split-Model project, complete these tasks:
```

== Definition Project Structure

After running the `make build` command, the policy's root directory contains the following directories and files shown in the directory tree:

----
├─ gcl.yaml # GCL file managed by the developer that describes the policy Definition
├─ exchange.json # Coordinates of the asset
├─ target/      # Contains the output of the policy definition build
├─ Makefile     # Provides all the callable actions required during the policy definition development lifecycle
├─ README.md   # Information regarding general help of available actions
└─.gitignore # Configuration for git Version Control system to ignore built assets
----

=== exchange.json

Contains the exchange asset coordinate of the policy definition.

=== target

Contains the policy asset definition published to Exchange. The directory and its content are included in the `.gitignore` file.

=== Definition Makefile

The Makefile contains all the callable actions, such as `make` commands, required for the policy development lifecycle. Execute all `make` commands at the root level of your custom policy directory:

* `make build`: Compiles the definition and generates the assets to be deployed.

> * `make publish`: Publishes the policy definition asset in Exchange. Since the publish goal is intended to publish a policy asset in development, the _assetId_ and name published will explicitly say `dev`, and the versions published will include a timestamp at the end of the version. For example:

Reviewer: ```suggestion
* `make publish`: Publishes the policy definition asset in Exchange. Since the publish goal is intended to publish a policy asset in development, the _assetId_ and name published explicitly say `dev`, and the versions published include a timestamp at the end of the version. For example:
```

** groupId: your configured organization id
** visible name: {Your policy name} Dev
** assetId: {your-policy-asset-id}-dev
** version: {your-policy-version}-20230618115723
+
For more information about publishing policies in Exchange, see xref:policies-pdk-publish-policies.adoc[].
* `make release`: Releases the policy definition asset in Exchange.
+
For more information about releasing policies in Exchange, see xref:policies-pdk-publish-policies.adoc[].

> * `make release-local`: Publishes the policy definition as a release asset to the local cache (no actual release to Exchange is made). You can override it. This target is useful if you are also developing the policy implementation.

Reviewer: ```suggestion
* `make release-local`: Publishes the policy definition as a release asset to the local cache (no actual release to Exchange is made). You can override it. This target is useful if you're also developing the policy implementation.
```

* `make help`: Lists all the `make` commands provided by PDK.

[[create-a-new-policy-implementation-project]]
== Create a New Policy Implementation Project

To create a custom policy implementation project, run the PDK CLI `create` command, with the `--project-mode implementation` flag, replacing `<my-custom-policy>` with the name of your new policy:

[source,ssh]
----
anypoint-cli-v4 pdk policy-project create --name <my-custom-policy> --project-mode implementation
----

Running the `create` command creates a custom policy project with the specified name. The project appears in a new directory with the same name as your policy with the `-implementation` suffix or a directory specified using the `--output-dir` flag.

> NOTE: If your implementation and definition projects have different names, link them by ensuring the `definition_asset_id.name` field in the implementation project's `Cargo.toml` file matches the asset ID in the definition project's `exchange.json` file.

Reviewer: ```suggestion
NOTE: If your implementation and definition projects have different names, link them by making sure the `definition_asset_id.name` field in the implementation project's `Cargo.toml` file matches the asset ID in the definition project's `exchange.json` file.
```

[[setup-the-pdk-build-environment]]
== Set Up the PDK Build Environment

After you <<create-a-new-policy-project, create a new policy project>>, you must run the `make setup` command to download additional software dependencies:

[source,ssh]
----
make setup
----

[[implementation-project-structure]]
== Implementation Project Structure

> After running the `make setup` command, the policy's root directory contains the following directories and files shown in the directory tree:

Reviewer: ```suggestion
After running the `make setup` command, the policy's root directory contains these directories and files shown in the directory tree:
```

----

=== target

Contains the files for the policy's Exchange asset implementation and the output of the project build. The target directory and its content are included in the `.gitignore` file.

=== Cargo.toml

Custom policies for Flex Gateway are developed in the Rust programming language and compiled to a WebAssembly binary.

The `Cargo.toml` is the Rust project manifest. Its `[package]` section contains the basic information about the policy implementation. The version specified in `Cargo.toml` matches the policy asset version. You can manage the policy version by editing the version number in `Cargo.toml`. The version follows the format `<major-version>.<minor-version>.<patch-version>`.

The toml also contains a section `[package.metadata.anypoint]` where you can specify:

* `group_id`: The policy group ID, which should match your Anypoint organization.
* `definition_asset_id`: The name of the policy definition and the version this implementation uses. If you need to work with a development version, append the `-DEV` (case sensitive) suffix, for example: `version = "1.0.0-DEV"`.

> * `implementation_asset_id`: The asset ID that will be used to release the policy.

Reviewer: ```suggestion
* `implementation_asset_id`: The asset ID that's used to release the policy.
```
